app/manager.py

```diff
@@ -1,127 +1,141 @@
+            for doc_id, details in row.docs.items():
+                if doc_id in res:
+                    res[doc_id]['score'] += details[0]
                 else:
-                    words_dict[word]['count'] += 1
-
-                if next_word:
-                    words_dict[word]['next'].add(fix_word(next_word))
+                    res[doc_id] = {}
+                    res[doc_id]['score'] = details[0]
+                    res[doc_id]['idx'] = []
 
-            counter += 1
+                res[doc_id]['idx'].append((row.word, details[1]))
 
-    for word in words_dict:
-        db_rep.upsert_word(word, doc_id, words_dict[word])
+        docs_rows = self._cassandra_repository.get_documents_by_ids(list(res.keys())).current_rows
+        docs_dict = {x.id: x for x in docs_rows}
+        res = {str((docs_dict[doc].title, docs_dict[doc].author)): res[doc] for doc in res}
 
+        return res
 
-def fix_word(word):
-    if word:
-        for char in bad_chars:
-            word = word.replace(char, '')
+    def exact(self, str_arg):
+        words = str_arg.split()
+        rows = self._cassandra_repository.get_rows_by_words(words, 'words')
+        res_docs = []
 
-    return word
+        if len(words) == len(rows.current_rows):
+            words_dict = {}
+            for row in rows.current_rows:
+                words_dict[row.word] = row.docs
 
+            for doc, details in words_dict[words[0]].items():
+                self.check_doc(words, words_dict, doc, details[2], 1, res_docs)
 
-def search(str_arg, case=False):
-    table = 'lower_words' if case else 'words'
+        docs_rows = self._cassandra_repository.get_documents_by_ids(res_docs).current_rows
 
-    res = {}
+        res = [str((doc.title, doc.author)) for doc in docs_rows]
 
-    rows = cassandra_repository.get_rows_by_words(str_arg.split(), table)
+        return str(res)
 
-    for row in rows.current_rows:
-        for doc_id, details in row.docs.items():
-            if doc_id in res:
-                res[doc_id]['score'] += details[0]
-            else:
-                res[doc_id] = {}
-                res[doc_id]['score'] = details[0]
-                res[doc_id]['idx'] = []
+    def check_doc(self, words, words_dict, doc, next_words, idx, res):
+        if idx == len(words):
+            res.append(doc)
+            return
 
-            res[doc_id]['idx'].append((row.word, details[1]))
-
-    docs_rows = cassandra_repository.get_documents_by_ids(list(res.keys())).current_rows
-    docs_dict = {x.id: x for x in docs_rows}
-    res = {str((docs_dict[doc].title, docs_dict[doc].author)): res[doc] for doc in res}
-
-    return res
-
-
-def exact(str_arg):
-    words = str_arg.split()
-    rows = db_rep.get_rows_by_words(words, 'words')
-    res_docs = []
-
-    if len(words) == len(rows.current_rows):
-        words_dict = {}
-        for row in rows.current_rows:
-            words_dict[row.word] = row.docs
+        if words[idx] in next_words:
```

Comment:
Not sure I like the next word approach - In the end you make a lot of effort to save all of them but only to save the last word query (you'll need to get the query anyway in order to continue).
But I like the recursion!

Reply:
All of the efforts are only in the indexing process. I thought of some more approaches for this but I found this one the most simple and efficient